plonky2/src/hash/keccak_sponge_sha256.rs

```diff
@@ -0,0 +1,72 @@
+use keccak::f1600 as keccak;
+use sha2::{Digest, Sha256};
+
+use crate::hash::hash_types::{BytesHash, RichField};
+use crate::hash::hashing::{PlonkyPermutation, SPONGE_WIDTH};
+use crate::plonk::config::Hasher;
+
+/// Keccak sponge pseudo-permutation (not necessarily one-to-one) used in the challenger.
```

Comment:
Hm I'm a little confused by this. Can you not use the Keccak permutation directly? We used a pseudo-permutation in `KeccakPermuation` for compatibility with EVM's `Keccak256`; do you have some constraint like that?

It doesn't seem like this is technically a sponge, since it absorbs all `SPONGE_WIDTH` inputs in the first round, rather than absorbing `SPONGE_RATE` inputs at a time.

Reply:
> Hm I'm a little confused by this. Can you not use the Keccak permutation directly? We used a pseudo-permutation in `KeccakPermuation` for compatibility with EVM's `Keccak256`; do you have some constraint like that?

How do we guarantee the 64-bit state elements fit in the field? Is it OK to just mod by `F::ORDER`? I was under the impression that would introduce some bias into the distribution, though I'm not sure by how much.

> It doesn't seem like this is technically a sponge, since it absorbs all `SPONGE_WIDTH` inputs in the first round, rather than absorbing `SPONGE_RATE` inputs at a time.

Thanks, that should be `SPONGE_RATE` not `SPONGE_WDITH` there.